lib/icingadb/icingadb.hpp

```diff
@@ -176,6 +183,15 @@ class IcingaDB : public ObjectImpl<IcingaDB>
 	Timer::Ptr m_StatsTimer;
 	WorkQueue m_WorkQueue{0, 1, LogNotice};
 
+	std::thread m_HistoryThread;
+	Bulker<RedisConnection::Query> m_HistoryBulker {4096, std::chrono::milliseconds(250)};
+
+	struct {
+		std::mutex Mutex;
+		std::condition_variable CV;
+		bool Value = false;
+	} m_HistoryThreadIsGoingToStop;
```

Comment:
I think you're basically building `std::promise<void>` yourself here.

You can also have a look at `std::async()` which allows running a function on a new thread and returns a `std::future` for the return value of that function.